### PR TITLE
Add new IPv6 types to docs where it's supported

### LINKED ISSUE #17993: Docs for IPv6
**Elasticsearch version**: master/5.0.0-alpha2

I notice we mention IPv4 several places that the new IPv6 type works:
~~https://www.elastic.co/guide/en/elasticsearch/reference/master/common-options.html#fuzziness~~
https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-types.html#_specialised_datatypes

### PR DESCRIPTION
**Pull request by eskibars**
Feature was completed in https://github.com/elastic/elasticsearch/pull/17746
This should close https://github.com/elastic/elasticsearch/issues/17993

### CODE REVIEW

File: docs/reference/api-conventions.asciidoc [hunk 1]
```diff
@@ -413,9 +413,9 @@ using the `fuzziness` parameter. The `fuzziness` parameter is context
 sensitive which means that it depends on the type of the field being queried:
 
 [float]
-==== Numeric, date and IPv4 fields
+==== Numeric, date and IP fields
 
-When querying numeric, date and IPv4 fields, `fuzziness` is interpreted as a
+When querying numeric, date, IPv4, and IPv6 fields, `fuzziness` is interpreted as a
```
Comment (clintongormley) [inline docs/reference/api-conventions.asciidoc]: Here I'd just use IP fields

**Comment by clintongormley** [comment]
I'd like @jpountz to confirm that fuzziness works as documented for the new IP fields before merging

**Comment by jpountz** [comment]
Good catch it does not. I can fix it but this makes me wonder whether we want it to work this way. I am wondering if it is useful at all since you would need to pass absurdly high values for it to be useful on ipv6 addresses: since there are 128 bits for addresses, you would need to pass 2<sup>64</sup> to match all ip addresses in the same `/64` for instance.

**Comment by clintongormley** [comment]
>  I can fix it but this makes me wonder whether we want it to work this way. I am wondering if it is useful at all since you would need to pass absurdly high values for it to be useful on ipv6 addresses: since there are 128 bits for addresses, you would need to pass 264 to match all ip addresses in the same /64 for instance.

Yeah.  I have no idea how much this is used, my gut says not a lot.  Possibly we should change the semantics for fuzzy on IP to say that (eg) `~16` behaves like netmask `/16`?  Would that make any sense?

**Comment by jpountz** [comment]
Given that term queries already support the netmask notation (if you run a term query on `::1/16`, it will match all ip addresses in the same `/16` as `::1` as expected), I'm not sure it would be useful. I'm wondering that we should just raise an exception if you run a fuzzy query on an ip field saying that it is not supported?

**Comment by clintongormley** [comment]
>  I'm wondering that we should just raise an exception if you run a fuzzy query on an ip field saying that it is not supported?
++

**Comment by eskibars** [comment]
I'm +1 to deprecate now and remove shortly, but -1 for removing support immediately for 5.0.  I don't have a good sense for how many people are using this style of query currently, but it's another thing that the deprecation logging we've been pushing in 2.3 won't be able to catch and would make upgrading harder.

**Comment by jpountz** [comment]
I opened #18016

**Comment by eskibars** [comment]
I pulled the fuzziness bit out to revert the docs there to IPv4 only

**Comment by rjernst** [comment]
> I'm +1 to deprecate now and remove shortly, but -1 for removing support immediately for 5.0. I don't have a good sense for how many people are using this style of query currently

Not removing esoteric things simply because the feature existed, but without any evidence that someone is using it, leads to complex legacy code that stays around for far longer than it should. We should remove this feature, it is only a burden on us and doesn't actually help users. It is very simple if there was someone using it to get an error that says "use a CIDR mask instead".

**Comment by eskibars** [comment]
> Not removing esoteric things simply because the feature existed, but without any evidence that someone is using it, leads to complex legacy code that stays around for far longer than it should. We should remove this feature, it is only a burden on us and doesn't actually help users. It is very simple if there was someone using it to get an error that says "use a CIDR mask instead".

Fundamentally, I agree with your statement as long as we have enough ways to derive that evidence as to whether a particular feature is being used or not and to what extent.  As of right now, I don't believe we do, and the fact that it's in the docs as a specific example of fuzzy leaves me especially cautious.  In lieu of any evidence either direction, I'm personally a proponent of airing on the side of "deprecate and then remove," carrying that code burden for a short time rather than flash removal.

**Comment by rjernst** [comment]
>  I'm personally a proponent of airing on the side of "deprecate and then remove," carrying that code burden for a short time rather than flash removal.

But we are not talking about a short time. We are talking about keeping it around for an entire major release, for something that shouldn't be a "feature" in the first place. Things like this hold back progress, because they require keeping apis in line with the "old" way of doing things. For example, we have tons of these methods of MappedFieldType that are really geared towards term dictionary based fields. It would be great to rework the field types so just text/keyword fields have them. But keeping this fuzzy query around for ip means holding back on any improvements there. It also means that *new* users will potentially fall into this trap of using this crazy "fuzzy" query on ip fields, instead of cidr masks like they should. And this last point is really important: we should lean on the side of removing things if it improves clarity of apis for new users.

**Comment by jpountz** [comment]
The argument that resonates most to me is that if anyone is using a fuzzy query on anything but a string/keyword/text field, it is more likely a mistake than on purpose. Since I cannot recall of any github issue or discuss post talking about this feature, I am leaning towards removing it.

**Comment by rmuir** [comment]
What is supposed to happen if a user uses the `~` operator on an ipv6 address? This isn't like ancient 1980s ipv4 where you could infer a "default" netmask or something.

Personally, I think we should focus on the real use cases (that is unarguably prefix query for network addresses). Means making sure it works in queryparsers/aggregations/wherever its supposed to work. 

Things like fuzzy/wildcards on ip addresses should just be an error message and we shouldn't spend valuable time on that, time that could be spent making ip addresses work well instead.

**Comment by eskibars** [comment]
These are great conversations and I think we should have them, but buried in a docs PR is probably not the best place for them.  I pulled the fuzzy update because we shouldn't have docs that don't reflect reality, whatever that reality happens to be.  If we're happy with what remains here, we can merge and then carry on the other conversations here outside of this PR

**Comment by jpountz** [comment]
+1 to merge the PR

**Pull request merged**